src/oid.rs

```diff
@@ -0,0 +1,245 @@
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 00 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 10 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 20 */
+ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, -1, -1, -1, -1, -1, -1, /* 30 */
+-1, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 40 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 50 */
+-1, 10, 11, 12, 13, 14, 15, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 60 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 70 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 80 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* 90 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* a0 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* b0 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* c0 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* d0 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* e0 */
+-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, /* f0 */
+];
+
+#[inline]
+fn fromhex(h: u8) -> i32 {
+    FROM_HEX[h as usize]
+}
+
+//static TO_HEX: [u8; 16] = [ '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'];
+static TO_HEX: &'static [u8; 16] = b"0123456789abcdef";
+
+#[inline]
+unsafe fn fmt_one(mut pstr: *mut u8, val: i32) -> *mut u8 {
+    *pstr = TO_HEX[(val >> 4) as usize];
+    pstr = pstr.offset(1);
+    *pstr = TO_HEX[(val & 0xf) as usize];
+    pstr.offset(1)
+}
+
+const GIT_OID_RAWSZ: usize = 20;
+const GIT_OID_HEXSZ: usize = 40;
+
+#[inline]
+fn oid_error_invalid(msg: &[u8]) -> Result<(), i32> {
+    unsafe {
+        giterr_set(git_error_t::GITERR_INVALID, b"Unable to parse OID - %s" as *const _, msg.as_ptr() as *const _);
+    }
+    Err(-1)
+}
+
+// git_oid
+#[derive(Default, PartialEq, Eq, PartialOrd)]
+#[repr(C)]
+pub struct Oid {
+    id: [u8; GIT_OID_RAWSZ],
+}
+
+impl Oid {
+
+    pub fn from_slice(out: &mut Oid, input: &[u8]) -> Result<(), i32> {
+        if input.len() == 0 {
+            return oid_error_invalid(b"too short");
+        }
+
+        if input.len() > GIT_OID_HEXSZ {
+            return oid_error_invalid(b"too long");
+        }
+
+        out.id = [0; 20];
+
+        for (p, c) in input.iter().enumerate() {
+            let v = fromhex(*c);
+            if v < 0 {
+                return oid_error_invalid(b"contains invalid characters");
+            }
+
+            // Apply alternatingly to the higher or lower nibble
+            let shift: u8 = ((!p & 0x1) * 4) as u8;
+            out.id[p/2] |= (v as u8) << shift;
+        }
+
+        Ok(())
+    }
+
+    fn cmp_slice(&self, s: &[u8]) -> Ordering {
+        let mut s_iter = s.iter();
+        for a in self.id.iter() {
+            let c1 = match s_iter.next() {
+                None => break,
+                Some(c) => *c,
```

Comment:
As mentioned above, for each byte in `self.id` we want two out of `s`; if we can do something like `for (a, b) in self.id.iter().zip(s_iter.group(2))` maybe we can use that as the loop (if it fills when the group isn't big enough).